From 5b80ae94824f720b8377d99d654016b52d5c5a27 Mon Sep 17 00:00:00 2001
From: Mathew Blowers <mathew.blowers@raspberrypi.com>
Date: Tue, 11 Aug 2026 15:56:23 +0100
Subject: [PATCH] flashchips: Add AT25EU0041A

Datasheet: https://www.renesas.com/en/document/dst/at25eu0041a-datasheet
Tested: probe, read, write and erase
Tested-with: Raspberry Pi 4
https://paste.flashrom.org/view.php?id=3886

Change-Id: If52e2935cc3bb3d8c62d558ce1898c594b6501c9
Signed-off-by: Mathew Blowers <mathew.blowers@raspberrypi.com>
Reviewed-on: https://review.coreboot.org/c/flashrom/+/94904
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Abdelkader Boudih <coreboot@seuros.com>
---
 flashchips/atmel.c   | 44 ++++++++++++++++++++++++++++++++++++++++++++
 include/flashchips.h |  1 +
 2 files changed, 45 insertions(+)

--- a/flashchips/atmel.c
+++ b/flashchips/atmel.c
@@ -514,6 +514,50 @@
 
 	{
 		.vendor		= "Atmel",
+		.name		= "AT25EU0041A",
+		.bustype	= BUS_SPI,
+		.manufacture_id	= ATMEL_ID,
+		.model_id	= ATMEL_AT25EU0041A,
+		.total_size	= 512,
+		.page_size	= 256,
+		.feature_bits	= FEATURE_WRSR_WREN | FEATURE_OTP,
+		.tested		= TEST_OK_PREW,
+		.probe		= PROBE_SPI_RDID,
+		.probe_timing	= TIMING_ZERO,
+		.block_erasers	=
+		{
+			{
+				.eraseblocks = { {256, 2048} },
+				.block_erase = SPI_BLOCK_ERASE_81,
+			}, {
+				.eraseblocks = { {256, 2048} },
+				.block_erase = SPI_BLOCK_ERASE_DB,
+			}, {
+				.eraseblocks = { {4 * 1024, 128} },
+				.block_erase = SPI_BLOCK_ERASE_20,
+			}, {
+				.eraseblocks = { {32 * 1024, 16} },
+				.block_erase = SPI_BLOCK_ERASE_52,
+			}, {
+				.eraseblocks = { {64 * 1024, 8} },
+				.block_erase = SPI_BLOCK_ERASE_D8,
+			}, {
+				.eraseblocks = { {512 * 1024, 1} },
+				.block_erase = SPI_BLOCK_ERASE_60,
+			}, {
+				.eraseblocks = { {512 * 1024, 1} },
+				.block_erase = SPI_BLOCK_ERASE_C7,
+			}
+		},
+		.printlock	= SPI_PRETTYPRINT_STATUS_REGISTER_BP4_SRWD,
+		.unlock		= SPI_DISABLE_BLOCKPROTECT_BP4_SRWD,
+		.write		= SPI_CHIP_WRITE256,
+		.read		= SPI_CHIP_READ,
+		.voltage	= {1650, 3600},
+	},
+
+	{
+		.vendor		= "Atmel",
 		/* The A suffix indicates 33MHz instead of 20MHz clock rate.
 		 * All other properties seem to be the same.*/
 		.name		= "AT25F1024(A)",
--- a/include/flashchips.h
+++ b/include/flashchips.h
@@ -131,6 +131,7 @@
 #define ATMEL_AT25DL161		0x4603	/* EDI 0x0100 */
 #define ATMEL_AT25DQ161		0x8600	/* EDI 0x0100 */
 #define ATMEL_AT25DQ321		0x8700	/* EDI 0x0100 */
+#define ATMEL_AT25EU0041A	0x1401
 #define ATMEL_AT25F512		0x60	/* Needs AT25F_RDID. ID from PCN and actual HW. Seems to be a relabeled AT25F1024. */
 #define ATMEL_AT25F512A		0x65	/* Needs AT25F_RDID */
 #define ATMEL_AT25F512B		0x6500
